File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.java
class Solution {
    public String reverseWords(String s) {
        char[] str = s.toCharArray();
        int start = 0;

        for(int end =0 ; end < str.length ; end++) {
            if(str[end] == ' ' || end == str.length - 1) {
                int left = start;
                int right = (str[end] == ' ') ? end - 1 : end;

                while(left < right) {
                    char temp = str[left];
                    str[left] = str[right];
                    str[right] = temp;

                    left++ ;
                    right-- ;
                }
                start = end+1;
            }
        }
        return new String(str);
    }
}

File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.py
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        str_list = list(s)
        start = 0
        length = len(str_list)

        for end in range(length):
            if str_list[end] == ' ' or end == length - 1:
                left = start
                right = (end - 1) if str_list[end] == ' ' else end

                while left < right:
                    temp = str_list[left]
                    str_list[left] = str_list[right]
                    str_list[right] = temp

                    left += 1
                    right -= 1
                
                start = end + 1

        return "".join(str_list)